### PR TITLE
useInbox and useOutbox: new awaitActivity method

### PR DESCRIPTION
**Pull request by srosset81**
- Introduce new `awaitActivity` that allows to watch for activities coming to the inbox or outbox.
- Refactor `useInbox` and `useOutbox` to use `useCollection` hook (the old fetch method didn't work anyway, and it relied on the SPARQL endpoint. In the future, we plan to add a filter option to collections)
- Convert these two hooks to TypeScript :1st_place_medal: 
- Add documentation

## Breaking changes :warning: 

- No `fetch` method is returned from `useInbox` and `useOutbox`. The first page of results is automatically fetched. More results can be fetched with `fetchNextPage`.
- The `DataProvider` type has been renamed to `SemanticDataProvider` to clarify the difference with React-Admin type.

### CODE REVIEW

**Comment by srosset81** [comment]
@mguihal @Laurin-W I will merge this tomorrow. If you have some inputs, I'd be happy to hear them.
FYI it's going to be used here https://github.com/activitypods/activitypods/pull/272

File: src/frontend/packages/activitypub-components/src/hooks/useAwaitActivity.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            reject(new Error(`${e.reason} (Code: ${e.code})`));
+          });
+
+          // If a list of activities is already loaded, verify if there is a match
+          if (existingActivities && checkExistingActivities) {
+            for (const a of existingActivities) {
+              if (typeof a !== 'string') {
+                if (matchActivity(a)) {
+                  webSocket.removeEventListener('message', onMessage);
+                  resolve(a);
```
Comment (Laurin-W) [inline src/frontend/packages/activitypub-components/src/hooks/useAwaitActivity.ts]: Maybe add a return statement here and at line 21?

File: src/frontend/packages/activitypub-components/src/hooks/useOutbox.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+          '@context': 'https://www.w3.org/ns/activitystreams',
+          ...activity
+        })
+      });
+      return headers.get('Location');
+    },
+    [url, dataProvider]
+  );
+
+  return { url, hasLiveUpdates, items, post, awaitActivity, owner: identity?.id, ...rest };
```
Comment (Laurin-W) [inline src/frontend/packages/activitypub-components/src/hooks/useOutbox.ts:43]: Could you add a small jsdoc description (awaitActivity, post, reference to useCollection hook)? This is incredibly useful when using those function in other apps, like the app boilerplate. And maybe a link to the semapps documentation :)

File: src/frontend/packages/semantic-data-provider/src/dataProvider/types.ts [hunk 3]
```diff
@@ -96,9 +96,10 @@ export type Configuration = {
   resources: Record<string, DataModel>;
 };
 
-export type DataProvider = RaDataProvider & {
+export type SemanticDataProvider = DataProvider & {
   getDataModels: () => Promise<Record<string, DataModel>>;
   getDataServers: () => Promise<DataServersConfig>;
+  fetch: (url: string, options?: HttpClientOptions) => ReturnType<typeof fetchUtils.fetchJson>;
```
Comment (Laurin-W) [inline src/frontend/packages/semantic-data-provider/src/dataProvider/types.ts:102]: 🙌

**Review by Laurin-W — COMMENTED** [review]
Looking good, I'm really looking forward to adding more type safety to the project! :D

**Review by mguihal — APPROVED** [review]

File: src/frontend/packages/activitypub-components/src/hooks/useOutbox.ts (hunk 2 above)
Comment (srosset81) [reply src/frontend/packages/activitypub-components/src/hooks/useOutbox.ts:43]: Done ! Thanks for the reminder.

**Pull request merged**